h2o-persist-gcs/src/main/java/water/persist/PersistGcs.java

```diff
@@ -0,0 +1,161 @@
+      final List<String> c = new ArrayList<>();
+      for (Bucket b : storage.list().iterateAll()) {
+        c.add(b.getName());
+      }
+      return c;
+    }
+  };
+
+  @Override
+  public List<String> calcTypeaheadMatches(String filter, int limit) {
+    final String input = GcsBlob.removePrefix(filter);
+    final String[] bk = input.split("/", 2);
+    if (bk.length == 1) {
+      return bucketCache.fetch(GcsBlob.KEY_PREFIX, bk[0], limit);
+    } else if (bk.length == 2) {
+      if (!keyCache.containsKey(bk[0])) {
+        StringCache sc = new StringCache() {
+          @Override
+          List<String> update() {
+            final List<String> cache = new ArrayList<>();
+            for (Blob b : storage.get(bk[0]).list().iterateAll()) {
+              cache.add(b.getName());
+            }
+            return cache;
+          }
+        };
+        keyCache.put(bk[0], sc);
+      }
+      return keyCache.get(bk[0]).fetch(GcsBlob.KEY_PREFIX + bk[0] + "/", bk[1], limit);
+    }
+    return new ArrayList<>(0);
+  }
+
+  @Override
+  public void importFiles(String path,
+                          String pattern,
+                          ArrayList<String> files,
+                          ArrayList<String> keys,
+                          ArrayList<String> fails,
+                          ArrayList<String> dels) {
+    final String bk[] = GcsBlob.removePrefix(path).split("/", 2);
+    final Bucket bucket = storage.get(bk[0]);
+    for (Blob blob : bucket.list().iterateAll()) {
+      if (bk.length == 1 || blob.getName().startsWith(bk[1])) {
+        final GcsBlob gcsBlob = GcsBlob.of(blob.getBlobId());
+        Log.info("Importing: " + gcsBlob.toString());
+        try {
+          final Key k = GcsFileVec.make(gcsBlob.getCanonical(), blob.getSize());
+          keys.add(k.toString());
+          files.add(gcsBlob.getCanonical());
+        } catch (Throwable t) {
+          fails.add(gcsBlob.getCanonical());
+        }
+      }
+    }
+  }
+
+  @Override
+  public InputStream open(final String path) {
+    final GcsBlob gcsBlob = GcsBlob.of(path);
+    Log.info("Openning: " + gcsBlob.toString());
+    final Blob blob = storage.get(gcsBlob.getBlobId());
+    return new InputStream() {
+      @Override
+      public int read() throws IOException {
+        // very naive version with reading byte by byte
+        int res = -1;
+        try (ReadChannel reader = blob.reader()) {
+          ByteBuffer bytes = ByteBuffer.allocate(1);
+          while (reader.read(bytes) > 0) {
+            res = bytes.getInt();
+          }
+        } catch (IOException e) {
+          throw new FSIOException(path, e);
+        }
+        return res;
+      }
+
+      @Override
+      public int read(byte b[], int off, int len) throws IOException {
+        if (b == null) {
+          throw new NullPointerException();
+        } else if (off < 0 || len < 0 || len > b.length - off) {
+          throw new IndexOutOfBoundsException();
```

Comment:
A message with reason and actual values would be a "nice-to-have" feature. It is easier to recognize what exactly happened from the logs, as there are multiple possibilities in the if statement.

Reply:
Resolved in new PR https://github.com/h2oai/h2o-3/pull/2119